### PR TITLE
Home Owners (Part 4) Allocations, Tenancy Type & More

### PR DESCRIPTION
**Pull request by dimak1**
*Issue #:* /bcgov/entity#12451

*Description of changes:*
- Total Allocations calculations
- Lowest Common Multiplier (aka common denominator) calculations
- Keep empty Groups (do not delete them and show an error instead)
- Home and Group Tenancy Types
- Home Owners Table error

### Screenshots

Screen Shot 2022-08-30 at 8 25 23 AM

Screen Shot 2022-08-30 at 8 26 48 AM

Screen Shot 2022-08-30 at 8 27 27 AM

Screen Shot 2022-08-30 at 8 29 23 AM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

**Review by ChasBoucher — APPROVED** [review]
Looks Good.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/FractionalOwnership.vue [hunk 1]
```diff
@@ -20,7 +20,7 @@
           :id="`fraction-amount-group-${groupId}`"
           label="Amount Owned by this Group"
           filled
-          v-model="fractionalData.interestNumerator"
+          v-model.number="fractionalData.interestNumerator"
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/FractionalOwnership.vue:23]: This is a good way to typecast 👍 
From the source of the value

Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/FractionalOwnership.vue:23]: Yes, we need numbers here, but our validation is not robust enough to validate this, so I had a workaround (for now) by using two `isNumber` rules.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnersTable.vue [hunk 2]
```diff
@@ -33,7 +32,7 @@
           </td>
         </tr>
 
-        <tr v-else :key="row.item.id" class="owner-info">
+        <tr v-else-if="row.item.id !== undefined" :key="row.item.id" class="owner-info">
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnersTable.vue]: Could this be simplified to `row.item.id` or `!!row.item.id` ?

File: ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnersTable.vue [hunk 3]
```diff
@@ -214,6 +236,11 @@ export default defineComponent({
     tbody > tr.v-row-group__header:hover {
       background: #e2e8ee !important;
     }
+
+    .no-owners-error {
+      text-align: center;
+      color: $error;
+    }
   }
```
Comment (cameron-eyds) [inline ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnersTable.vue:244]: I'm not sure you need a unique error class here, `.error-text` in the base.scss handles the color and there is a vuetify helper class `text-center` to help with alignment.

File: ppr-ui/src/composables/mhrRegistration/home-owner.ts [hunk 4]
```diff
@@ -44,6 +49,48 @@ export function useHomeOwners (isPerson: boolean = false, isEditMode: boolean =
...
+      return HomeTenancyTypes.NA
+    }
+  }
+
+  /**
+   * Get Ownership Allocation status object to conveniently show total allocation
+   * and an allocation error status if exists
+   */
+  const getTotalOwnershipAllocationStatus = (): MhrRegistrationTotalOwnershipAllocationIF => {
+    const totalFractionalNominator = sumBy(getMhrRegistrationHomeOwnerGroups.value, 'interestNumerator')
```
Comment (cameron-eyds) [inline ppr-ui/src/composables/mhrRegistration/home-owner.ts:86]: Other developers could look up `sumBy` in Lodash but perhaps a small comment here or in the docblock would be helpful in describing whats happening here in particular.

File: ppr-ui/src/composables/mhrRegistration/home-owner.ts [hunk 5]
```diff
@@ -180,14 +203,75 @@ export function useHomeOwners (isPerson: boolean = false, isEditMode: boolean =
...
+    })
+
+    const updatedFractionalData = currentFractionalData
+
+    // Update current fractional data with new values
+    updatedFractionalData.interestNumerator =
+      (LCM / currentFractionalData.interestTotal) * currentFractionalData.interestNumerator
+    updatedFractionalData.interestTotal = LCM
+
+    return { updatedGroups, updatedFractionalData }
```
Comment (cameron-eyds) [inline ppr-ui/src/composables/mhrRegistration/home-owner.ts]: Some complexity for sure, thank you the additional comments and examples. 100% required for other developers 👍

File: ppr-ui/src/views/newMhrRegistration/HomeOwners.vue [hunk 6]
```diff
@@ -227,4 +234,9 @@ export default defineComponent({
     vertical-align: text-bottom;
   }
 }
+.fractional-error-msg {
+  font-size: 14px;
+  line-height: 16px;
+  color: $error;
+}
```
Comment (cameron-eyds) [inline ppr-ui/src/views/newMhrRegistration/HomeOwners.vue]: Same as above, not sure a unique class is required.

**Review by cameron-eyds — APPROVED** [review]
Nice job! 
Great work on the comments/examples to help with the complexities 👍 
Just some very minor comments.

File: ppr-ui/src/composables/mhrRegistration/home-owner.ts (hunk 5 above)
Comment (dimak1) [reply ppr-ui/src/composables/mhrRegistration/home-owner.ts]: Yea, I tried to look up fractions math library but couldn't find exactly what we needed, so had to do it manually.

File: ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnersTable.vue (hunk 2 above)
Comment (dimak1) [reply ppr-ui/src/components/mhrRegistration/HomeOwners/HomeOwnersTable.vue]: Good catch!

File: ppr-ui/src/composables/mhrRegistration/home-owner.ts (hunk 5 above)
Comment (dimak1) [reply ppr-ui/src/composables/mhrRegistration/home-owner.ts]: Basically, for fractions, to find the lowest common denominator, we need to multiply all denominators (from all groups), and then based on that value update each nominator (for all groups). I can add that in as well. 
Thanks for the feedback!

File: ppr-ui/src/composables/mhrRegistration/home-owner.ts (hunk 4 above)
Comment (dimak1) [reply ppr-ui/src/composables/mhrRegistration/home-owner.ts:86]: Yea sounds good, I will add a comment.

**Review by doug-lovett — APPROVED** [review]
Nice, quite complicated.

**Comment by dimak1** [comment]
@cameron-freshworks addressed comments, found a bug and fixed it.

**Review by ChasBoucher — APPROVED** [review]
Looks Good...

**Pull request merged**